File: 백준/Gold/11000. 강의실 배정/강의실 배정.py
import sys
input = sys.stdin.readline

## 우선 순위 큐 이용해 정렬 상태 유지
from heapq import heappush, heappop

arr = []
heap = []
n = int(input())

for i in range(n):
    start, end = map(int, input().split())
    arr.append((start, end))

arr.sort() # 시작 시간 기준 정렬
heappush(heap, arr[0][1]) # 첫 회의 종료 시간을 우선순위 큐

for i in range(1,n):
    if (arr[i][0]<heap[0]): 
        # 그 다음에 있는 회의가 이전 회의보다 일찍 시작하면 새로운 회의실 사용
        heappush(heap, arr[i][1])
    else:
        # 그 다음에 있는 회의가 이전 회의 종료보다 늦게 시작하면 기존 회의실 사용
        heappop(heap)
        heappush(heap,arr[i][1])


print(len(heap))

File: 백준/Gold/11000. 강의실 배정/강의실 배정.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];

rl.on("line", (line) => {
  lines.push(line.trim());
}).on("close", () => {
  const N = Number(lines[0]);
  let lectures = lines.slice(1).map((e) => e.split(" ").map(Number));
  const schedules = [];

  // 강의 시작과 종료를 별도로 저장
  lectures.forEach((lecture) => {
    schedules.push({ time: lecture[0], start: 1 });
    schedules.push({ time: lecture[1], start: -1 });
  });

  // 정렬: 시간이 같으면 종료(-1)가 먼저 오도록
  schedules.sort((a, b) =>
    a.time === b.time ? a.start - b.start : a.time - b.time
  );

  let cnt = 0;
  let answer = 0;

  schedules.forEach((schedule) => {
    if (schedule.start === 1) {
      cnt++; // 강의 시작 -> 강의실 사용 증가
    } else {
      cnt--; // 강의 종료 -> 강의실 사용 감소
    }
    answer = Math.max(answer, cnt); // 최댓값 갱신
  });

  console.log(answer);
});
